BlueprintUI/Sources/Element/ElementState.swift

```diff
@@ -0,0 +1,597 @@
+    /// The method returns the output from your `toTrack` callback, as well as
+    /// an `Environment.Subset` that represents the measurement or layout's dependencies on the `Environment`.
+    ///
+    /// If there are no dependencies (`toTrack` did not read from the `Environment`), no subset is returned.
+    ///
+    /// Additionally, if the element is not comparable, no subset is returned or tracked.
+    private func trackEnvironmentReads<Output>(
+        for layoutPass: Environment.LayoutPass,
+        with environment: Environment,
+        in toTrack: (Environment) -> Output
+    ) -> (Output, Environment.Subset?) {
+
+        switch layoutPass {
+        case .measurement:
+            if comparability.isComparable {
+                break
+            } else {
+                return (toTrack(environment), nil)
+            }
+        case .layout:
+            if comparability == .comparableElement {
+                break
+            } else {
+                return (toTrack(environment), nil)
+            }
+        }
+
+        var environment = environment
+        var observedKeys = Set<Environment.StorageKey>()
+
+        environment.subscribeToReads(for: layoutPass) { key in
+            observedKeys.insert(key)
+        }
+
+        let output = toTrack(environment)
+
+        return (output, environment.subset(with: observedKeys))
+    }
+
+    /// Creates a new child `ElementState` for the provided `Element`,
+    /// or returns an existing one if it already exists.
+    ///
+    /// The first time the element is seen during a layout traversal, we will call `update`
+    /// on its backing `ElementState` to keep internal state in sync.
+    func childState(
+        for child: Element,
+        in environment: Environment,
+        with identifier: ElementIdentifier
+    ) -> ElementState {
+
+        if let existing = children[identifier] {
+
+            if existing.wasVisited == false {
+                existing.update(with: child, in: environment, identifier: identifier)
+            }
+
+            existing.wasVisited = true
+
+            return existing
+        } else {
+            let new = ElementState(
+                parent: self,
+                identifier: identifier,
+                element: child,
+                signpostRef: signpostRef,
+                name: name
+            )
+
+            children[identifier] = new
+
+            return new
+        }
+    }
+
+    /// To be called at the beginning of the layout cycle by the owner
+    /// of the `ElementStateTree`, to set up the tree for a traversal.
+    func prepareForLayout() {
+        recursiveForEach {
+            $0.wasVisited = false
+            $0.wasUpdateEquivalent = false
+        }
+    }
+
+    /// To be called at the end of the layout cycle by the owner
+    /// of the `ElementStateTree`, to tear down any old state,
+    /// or throw out caches which should not be maintained across layout cycles.
+    func finishedLayout() {
```

Comment:
Nit:
```suggestion
    func cleanupAfterLayout() {
```
?

Reply:
I am going to leave this for now; since I think I can make both prepare and finish a single function call; will try it in https://github.com/square/Blueprint/pull/389.